File: app/src/main/java/com/amaury/pointage/v2/engine/PayrollEngineV2.kt
package com.amaury.pointage.v2.engine

import com.amaury.pointage.v2.model.*

data class OvertimeTierV2(val fromMinutes:Int,val toMinutes:Int?,val multiplier:Double)

data class PayrollRulesV2(
    /** Seuil hebdomadaire confirmé par contrat/règle. Null = inconnu. */
    val weeklyRegularMinutes:Int? = null,
    /** Aucun palier n'est inventé : ils doivent venir d'une règle confirmée. */
    val overtimeTiers:List<OvertimeTierV2> = emptyList(),
    val nightMultiplier:Double? = null,
    val saturdayMultiplier:Double? = null,
    val sundayMultiplier:Double? = null
)

data class PayrollWeekV2(val paidMinutes:Int,val nightMinutes:Int=0,val saturdayMinutes:Int=0,val sundayMinutes:Int=0)

data class PayrollResultV2(
    val regularGross:Double,
    val overtimeGross:Double,
    val premiumsGross:Double,
    val fixedPremiumsGross:Double,
    val baskets:Double,
    val grossEstimate:Double,
    val deductions:Double,
    val netBeforeUnknownContributions:Double,
    val traces:List<String>
)

/**
 * Calcul déterministe : aucune durée ou majoration n'est supposée.
 * Le seuil hebdomadaire vient d'abord d'une règle confirmée, sinon du contrat.
 * Les majorations restent à zéro tant qu'aucun palier n'a été fourni.
 */
object PayrollEngineV2 {
    fun calculate(
        contract:ContractV2,
        weeks:List<PayrollWeekV2>,
        rules:PayrollRulesV2,
        premiums:List<PremiumV2> = emptyList(),
        baskets:List<BasketV2> = emptyList(),
        deductions:List<DeductionV2> = emptyList()
    ):PayrollResultV2 {
        if (contract.type == ContractTypeV2.FORFAIT_DAYS || contract.type == ContractTypeV2.FORFAIT_HOURS) {
            return calculateForfait(contract, premiums, baskets, deductions)
        }

        val rate = requireNotNull(contract.grossHourlyRate) { "Taux horaire brut obligatoire" }
        require(rate > 0.0) { "Taux horaire brut invalide" }

        val regularLimit = rules.weeklyRegularMinutes
            ?: contract.contractualWeeklyMinutes
            ?: error("Durée hebdomadaire contractuelle/règle obligatoire")
        require(regularLimit > 0) { "Durée hebdomadaire invalide" }

        var regularMinutes = 0
        var overtimeGross = 0.0
        var extras = 0.0
        val trace = mutableListOf<String>()

        weeks.forEach { week ->
            val paid = week.paidMinutes.coerceAtLeast(0)
            val regular = minOf(paid, regularLimit)
            regularMinutes += regular

            rules.overtimeTiers.forEach { tier ->
                require(tier.fromMinutes >= regularLimit) { "Palier d'heures supplémentaires incohérent" }
                require(tier.multiplier >= 1.0) { "Multiplicateur d'heures supplémentaires invalide" }
                val end = tier.toMinutes ?: Int.MAX_VALUE
                val minutes = (minOf(paid, end) - maxOf(regularLimit, tier.fromMinutes)).coerceAtLeast(0)
                if (minutes > 0) overtimeGross += minutes / 60.0 * rate * tier.multiplier
            }

            rules.nightMultiplier?.let { multiplier ->
                require(multiplier >= 1.0)
                if (week.nightMinutes > 0) extras += week.nightMinutes / 60.0 * rate * (multiplier - 1.0)
            }
            rules.saturdayMultiplier?.let { multiplier ->
                require(multiplier >= 1.0)
                if (week.saturdayMinutes > 0) extras += week.saturdayMinutes / 60.0 * rate * (multiplier - 1.0)
            }
            rules.sundayMultiplier?.let { multiplier ->
                require(multiplier >= 1.0)
                if (week.sundayMinutes > 0) extras += week.sundayMinutes / 60.0 * rate * (multiplier - 1.0)
            }
        }

        val regularGross = regularMinutes / 60.0 * rate
        val fixed = premiums.sumOf { it.amount }
        val basketTotal = baskets.sumOf { it.amount }
        val gross = regularGross + overtimeGross + extras + fixed
        val deductionsTotal = deductions.sumOf { it.amount }.coerceAtLeast(0.0)

        trace += "Temps payé V2 + durée contractuelle/règles confirmées"
        if (rules.overtimeTiers.isEmpty()) trace += "Aucune majoration d'heures supplémentaires appliquée : règle non fournie"
        if (baskets.isNotEmpty()) trace += "Paniers suivis séparément du brut estimé"

        return PayrollResultV2(
            regularGross,
            overtimeGross,
            extras,
            fixed,
            basketTotal,
            gross,
            deductionsTotal,
            (gross - deductionsTotal).coerceAtLeast(0.0),
            trace
        )
    }

    private fun calculateForfait(
        contract:ContractV2,
        premiums:List<PremiumV2>,
        baskets:List<BasketV2>,
        deductions:List<DeductionV2>
    ):PayrollResultV2 {
        val monthlyGross = requireNotNull(contract.monthlyGrossSalary) {
            "Salaire brut mensuel convenu obligatoire pour un forfait"
        }
        require(monthlyGross > 0.0) { "Salaire brut mensuel convenu invalide" }

        when (contract.type) {
            ContractTypeV2.FORFAIT_HOURS -> {
                requireNotNull(contract.forfaitHoursPeriod) { "Période du forfait heures obligatoire" }
                val hours = requireNotNull(contract.forfaitHours) { "Nombre d'heures du forfait obligatoire" }
                require(hours > 0.0) { "Nombre d'heures du forfait invalide" }
            }
            ContractTypeV2.FORFAIT_DAYS -> {
                val days = requireNotNull(contract.forfaitAnnualDays) { "Nombre annuel de jours obligatoire" }
                require(days > 0.0 && days <= 218.0) { "Nombre annuel de jours du forfait invalide" }
            }
            else -> error("Type de forfait incohérent")
        }

        val fixed = premiums.sumOf { it.amount }
        val basketTotal = baskets.sumOf { it.amount }
        val gross = monthlyGross + fixed
        val deductionsTotal = deductions.sumOf { it.amount }.coerceAtLeast(0.0)
        val trace = mutableListOf<String>()
        trace += when (contract.type) {
            ContractTypeV2.FORFAIT_HOURS -> "Forfait heures : salaire brut mensuel convenu utilisé comme base ; les heures du forfait ne sont pas reconverties artificiellement en taux horaire."
            ContractTypeV2.FORFAIT_DAYS -> "Forfait jours : salaire brut mensuel convenu utilisé comme base ; aucun taux horaire n'est inventé."
            else -> error("Type de forfait incohérent")
        }
        trace += "Les pointages servent au suivi du temps/charge et ne recalculent pas automatiquement la rémunération forfaitaire."
        if (baskets.isNotEmpty()) trace += "Paniers suivis séparément du brut estimé"

        return PayrollResultV2(
            regularGross = monthlyGross,
            overtimeGross = 0.0,
            premiumsGross = 0.0,
            fixedPremiumsGross = fixed,
            baskets = basketTotal,
            grossEstimate = gross,
            deductions = deductionsTotal,
            netBeforeUnknownContributions = (gross - deductionsTotal).coerceAtLeast(0.0),
            traces = trace
        )
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/PayrollEngineV2.swift
import Foundation

enum ContractTypeV2: String, Codable, Equatable {
    case fullTime = "FULL_TIME"
    case partTime = "PART_TIME"
    case forfaitHours = "FORFAIT_HOURS"
    case forfaitDays = "FORFAIT_DAYS"
    case forfait = "FORFAIT"
    case other = "OTHER"
}

enum ForfaitHoursPeriodV2: String, Codable, Equatable {
    case week = "WEEK"
    case month = "MONTH"
    case year = "YEAR"
}

enum PeriodicityV2: String, Codable, Equatable {
    case oneOff = "ONE_OFF"
    case daily = "DAILY"
    case weekly = "WEEKLY"
    case monthly = "MONTHLY"
    case yearly = "YEARLY"
}

struct ContractV2: Equatable {
    let id: String
    let employerId: String
    let type: ContractTypeV2
    let contractualWeeklyMinutes: Int?
    let grossHourlyRate: Double?
    let hireDateEpochDay: Int64?
    let payrollCutoffDay: Int?
    let forfaitHoursPeriod: ForfaitHoursPeriodV2?
    let forfaitHours: Double?
    let forfaitAnnualDays: Double?
    let monthlyGrossSalary: Double?

    init(
        id: String,
        employerId: String,
        type: ContractTypeV2,
        contractualWeeklyMinutes: Int?,
        grossHourlyRate: Double?,
        hireDateEpochDay: Int64?,
        payrollCutoffDay: Int? = nil,
        forfaitHoursPeriod: ForfaitHoursPeriodV2? = nil,
        forfaitHours: Double? = nil,
        forfaitAnnualDays: Double? = nil,
        monthlyGrossSalary: Double? = nil
    ) {
        self.id = id
        self.employerId = employerId
        self.type = type
        self.contractualWeeklyMinutes = contractualWeeklyMinutes
        self.grossHourlyRate = grossHourlyRate
        self.hireDateEpochDay = hireDateEpochDay
        self.payrollCutoffDay = payrollCutoffDay
        self.forfaitHoursPeriod = forfaitHoursPeriod
        self.forfaitHours = forfaitHours
        self.forfaitAnnualDays = forfaitAnnualDays
        self.monthlyGrossSalary = monthlyGrossSalary
    }
}

struct PremiumV2: Equatable {
    let id: String
    let label: String
    let amount: Double
    let periodicity: PeriodicityV2
    let taxable: Bool

    init(id: String, label: String, amount: Double, periodicity: PeriodicityV2, taxable: Bool = true) {
        self.id = id
        self.label = label
        self.amount = amount
        self.periodicity = periodicity
        self.taxable = taxable
    }
}

struct BasketV2: Equatable {
    let id: String
    let label: String
    let amount: Double
    let night: Bool

    init(id: String, label: String, amount: Double, night: Bool = false) {
        self.id = id
        self.label = label
        self.amount = amount
        self.night = night
    }
}

struct DeductionV2: Equatable {
    let id: String
    let label: String
    let amount: Double
    let recurring: Bool
}

struct OvertimeTierV2: Equatable {
    let fromMinutes: Int
    let toMinutes: Int?
    let multiplier: Double
}

struct PayrollRulesV2: Equatable {
    let weeklyRegularMinutes: Int?
    let overtimeTiers: [OvertimeTierV2]
    let nightMultiplier: Double?
    let saturdayMultiplier: Double?
    let sundayMultiplier: Double?
    let publicHolidayMultiplier: Double?

    init(
        weeklyRegularMinutes: Int? = nil,
        overtimeTiers: [OvertimeTierV2] = [],
        nightMultiplier: Double? = nil,
        saturdayMultiplier: Double? = nil,
        sundayMultiplier: Double? = nil,
        publicHolidayMultiplier: Double? = nil
    ) {
        self.weeklyRegularMinutes = weeklyRegularMinutes
        self.overtimeTiers = overtimeTiers
        self.nightMultiplier = nightMultiplier
        self.saturdayMultiplier = saturdayMultiplier
        self.sundayMultiplier = sundayMultiplier
        self.publicHolidayMultiplier = publicHolidayMultiplier
    }
}

struct PayrollWeekV2: Equatable {
    let paidMinutes: Int
    let nightMinutes: Int
    let saturdayMinutes: Int
    let sundayMinutes: Int
    let publicHolidayMinutes: Int

    init(
        paidMinutes: Int,
        nightMinutes: Int = 0,
        saturdayMinutes: Int = 0,
        sundayMinutes: Int = 0,
        publicHolidayMinutes: Int = 0
    ) {
        self.paidMinutes = paidMinutes
        self.nightMinutes = nightMinutes
        self.saturdayMinutes = saturdayMinutes
        self.sundayMinutes = sundayMinutes
        self.publicHolidayMinutes = publicHolidayMinutes
    }
}

struct PayrollResultV2: Equatable {
    let regularGross: Double
    let overtimeGross: Double
    let premiumsGross: Double
    let fixedPremiumsGross: Double
    let baskets: Double
    let grossEstimate: Double
    let deductions: Double
    let netBeforeUnknownContributions: Double
    let complementaryMinutes: Int
    let grossReliable: Bool
    let traces: [String]
}

enum PayrollEngineErrorV2: Error, Equatable {
    case missingHourlyRate
    case invalidHourlyRate
    case missingWeeklyDuration
    case invalidWeeklyDuration
    case invalidOvertimeTier
    case invalidMultiplier
    case missingMonthlyGross
    case invalidMonthlyGross
    case missingForfaitHoursPeriod
    case missingForfaitHours
    case invalidForfaitHours
    case missingForfaitAnnualDays
    case invalidForfaitAnnualDays
    case incoherentForfaitType
}

/// Miroir Swift du moteur brut PayrollEngineV2 Android.
///
/// Le moteur ne déduit aucune règle juridique : il ne consomme que des durées,
/// paliers et multiplicateurs déjà confirmés par les couches amont. Pour le temps partiel,
/// tant qu'aucune stipulation conventionnelle structurée plus précise n'est fournie,
/// le barème supplétif des heures complémentaires reste estimatif et rend le brut non fiable.
enum PayrollEngineV2 {
    static func calculate(
        contract: ContractV2,
        weeks: [PayrollWeekV2],
        rules: PayrollRulesV2,
        premiums: [PremiumV2] = [],
        baskets: [BasketV2] = [],
        deductions: [DeductionV2] = []
    ) throws -> PayrollResultV2 {
        if contract.type == .forfaitDays || contract.type == .forfaitHours {
            return try calculateForfait(
                contract: contract,
                premiums: premiums,
                baskets: baskets,
                deductions: deductions
            )
        }

        guard let rate = contract.grossHourlyRate else { throw PayrollEngineErrorV2.missingHourlyRate }
        guard rate > 0, rate.isFinite else { throw PayrollEngineErrorV2.invalidHourlyRate }

        guard let regularLimit = rules.weeklyRegularMinutes ?? contract.contractualWeeklyMinutes else {
            throw PayrollEngineErrorV2.missingWeeklyDuration
        }
        guard regularLimit > 0 else { throw PayrollEngineErrorV2.invalidWeeklyDuration }

        if contract.type == .partTime {
            return try calculatePartTime(
                contract: contract,
                weeks: weeks,
                rules: rules,
                rate: rate,
                regularLimit: regularLimit,
                premiums: premiums,
                baskets: baskets,
                deductions: deductions
            )
        }

        var regularMinutes = 0
        var overtimeGross = 0.0
        var extras = 0.0
        var traces: [String] = []

        for week in weeks {
            let paid = max(0, week.paidMinutes)
            regularMinutes += min(paid, regularLimit)

            for tier in rules.overtimeTiers {
                guard tier.fromMinutes >= regularLimit else { throw PayrollEngineErrorV2.invalidOvertimeTier }
                guard tier.multiplier >= 1, tier.multiplier.isFinite else {
                    throw PayrollEngineErrorV2.invalidMultiplier
                }
                let end = tier.toMinutes ?? Int.max
                let minutes = max(0, min(paid, end) - max(regularLimit, tier.fromMinutes))
                if minutes > 0 {
                    overtimeGross += Double(minutes) / 60.0 * rate * tier.multiplier
                }
            }

            extras += try premiumExtras(for: week, rate: rate, rules: rules)
        }

        let regularGross = Double(regularMinutes) / 60.0 * rate
        let fixed = premiums.reduce(0.0) { $0 + $1.amount }
        let basketTotal = baskets.reduce(0.0) { $0 + $1.amount }
        let gross = regularGross + overtimeGross + extras + fixed
        let deductionsTotal = max(0, deductions.reduce(0.0) { $0 + $1.amount })

        traces.append("Temps payé V2 + durée contractuelle/règles confirmées")
        if rules.overtimeTiers.isEmpty {
            traces.append("Aucune majoration d'heures supplémentaires appliquée : règle non fournie")
        }
        if !baskets.isEmpty {
            traces.append("Paniers suivis séparément du brut estimé")
        }

        return PayrollResultV2(
            regularGross: regularGross,
            overtimeGross: overtimeGross,
            premiumsGross: extras,
            fixedPremiumsGross: fixed,
            baskets: basketTotal,
            grossEstimate: gross,
            deductions: deductionsTotal,
            netBeforeUnknownContributions: max(0, gross - deductionsTotal),
            complementaryMinutes: 0,
            grossReliable: true,
            traces: traces
        )
    }

    private static func calculatePartTime(
        contract: ContractV2,
        weeks: [PayrollWeekV2],
        rules: PayrollRulesV2,
        rate: Double,
        regularLimit: Int,
        premiums: [PremiumV2],
        baskets: [BasketV2],
        deductions: [DeductionV2]
    ) throws -> PayrollResultV2 {
        guard let contractualWeeklyMinutes = contract.contractualWeeklyMinutes,
              contractualWeeklyMinutes > 0 else {
            throw PayrollEngineErrorV2.missingWeeklyDuration
        }

        var complementaryMinutes = 0
        var complementaryGross = 0.0
        var extras = 0.0
        var traces: [String] = []

        for week in weeks {
            let complementary = try PartTimeComplementaryHoursV2.calculateWeek(
                contractualMinutes: contractualWeeklyMinutes,
                paidMinutes: week.paidMinutes,
                grossHourlyRate: rate
            )
            complementaryMinutes += complementary.complementaryMinutes
            complementaryGross += complementary.grossToAdd
            traces.append(contentsOf: complementary.warnings)
            extras += try premiumExtras(for: week, rate: rate, rules: rules)
        }

        let monthlyBaseMinutes = Double(contractualWeeklyMinutes) * 52.0 / 12.0
        let regularGross = monthlyBaseMinutes / 60.0 * rate
        let fixed = premiums.reduce(0.0) { $0 + $1.amount }
        let basketTotal = baskets.reduce(0.0) { $0 + $1.amount }
        let gross = regularGross + complementaryGross + extras + fixed
        let deductionsTotal = max(0, deductions.reduce(0.0) { $0 + $1.amount })
        let provisionalComplementaryRateUsed = complementaryMinutes > 0

        traces.append("Salaire de base mensualisé temps partiel : durée contractuelle × 52/12 × taux horaire.")
        if provisionalComplementaryRateUsed {
            traces.append("Temps partiel : barème supplétif des heures complémentaires appliqué (+10 % puis +25 %) ; brut à confirmer tant qu'aucune stipulation conventionnelle structurée plus précise n'est intégrée.")
        }
        if !baskets.isEmpty {
            traces.append("Paniers suivis séparément du brut estimé")
        }

        return PayrollResultV2(
            regularGross: regularGross,
            overtimeGross: complementaryGross,
            premiumsGross: extras,
            fixedPremiumsGross: fixed,
            baskets: basketTotal,
            grossEstimate: gross,
            deductions: deductionsTotal,
            netBeforeUnknownContributions: max(0, gross - deductionsTotal),
            complementaryMinutes: complementaryMinutes,
            grossReliable: !provisionalComplementaryRateUsed,
            traces: unique(traces)
        )
    }

    private static func calculateForfait(
        contract: ContractV2,
        premiums: [PremiumV2],
        baskets: [BasketV2],
        deductions: [DeductionV2]
    ) throws -> PayrollResultV2 {
        guard let monthlyGross = contract.monthlyGrossSalary else {
            throw PayrollEngineErrorV2.missingMonthlyGross
        }
        guard monthlyGross > 0, monthlyGross.isFinite else {
            throw PayrollEngineErrorV2.invalidMonthlyGross
        }

        switch contract.type {
        case .forfaitHours:
            guard contract.forfaitHoursPeriod != nil else {
                throw PayrollEngineErrorV2.missingForfaitHoursPeriod
            }
            guard let hours = contract.forfaitHours else {
                throw PayrollEngineErrorV2.missingForfaitHours
            }
            guard hours > 0, hours.isFinite else {
                throw PayrollEngineErrorV2.invalidForfaitHours
            }
        case .forfaitDays:
            guard let days = contract.forfaitAnnualDays else {
                throw PayrollEngineErrorV2.missingForfaitAnnualDays
            }
            guard days > 0, days <= 218, days.isFinite else {
                throw PayrollEngineErrorV2.invalidForfaitAnnualDays
            }
        default:
            throw PayrollEngineErrorV2.incoherentForfaitType
        }

        let fixed = premiums.reduce(0.0) { $0 + $1.amount }
        let basketTotal = baskets.reduce(0.0) { $0 + $1.amount }
        let gross = monthlyGross + fixed
        let deductionsTotal = max(0, deductions.reduce(0.0) { $0 + $1.amount })
        var traces: [String] = []

        switch contract.type {
        case .forfaitHours:
            traces.append("Forfait heures : salaire brut mensuel convenu utilisé comme base ; les heures du forfait ne sont pas reconverties artificiellement en taux horaire.")
        case .forfaitDays:
            traces.append("Forfait jours : salaire brut mensuel convenu utilisé comme base ; aucun taux horaire n'est inventé.")
        default:
            throw PayrollEngineErrorV2.incoherentForfaitType
        }
        traces.append("Les pointages servent au suivi du temps/charge et ne recalculent pas automatiquement la rémunération forfaitaire.")
        if !baskets.isEmpty {
            traces.append("Paniers suivis séparément du brut estimé")
        }

        return PayrollResultV2(
            regularGross: monthlyGross,
            overtimeGross: 0,
            premiumsGross: 0,
            fixedPremiumsGross: fixed,
            baskets: basketTotal,
            grossEstimate: gross,
            deductions: deductionsTotal,
            netBeforeUnknownContributions: max(0, gross - deductionsTotal),
            complementaryMinutes: 0,
            grossReliable: true,
            traces: traces
        )
    }

    private static func premiumExtras(
        for week: PayrollWeekV2,
        rate: Double,
        rules: PayrollRulesV2
    ) throws -> Double {
        var extras = 0.0
        if let multiplier = rules.nightMultiplier {
            try validateMultiplier(multiplier)
            if week.nightMinutes > 0 {
                extras += Double(week.nightMinutes) / 60.0 * rate * (multiplier - 1.0)
            }
        }
        if let multiplier = rules.saturdayMultiplier {
            try validateMultiplier(multiplier)
            if week.saturdayMinutes > 0 {
                extras += Double(week.saturdayMinutes) / 60.0 * rate * (multiplier - 1.0)
            }
        }
        if let multiplier = rules.sundayMultiplier {
            try validateMultiplier(multiplier)
            if week.sundayMinutes > 0 {
                extras += Double(week.sundayMinutes) / 60.0 * rate * (multiplier - 1.0)
            }
        }
        if let multiplier = rules.publicHolidayMultiplier {
            try validateMultiplier(multiplier)
            if week.publicHolidayMinutes > 0 {
                extras += Double(week.publicHolidayMinutes) / 60.0 * rate * (multiplier - 1.0)
            }
        }
        return extras
    }

    private static func validateMultiplier(_ multiplier: Double) throws {
        guard multiplier >= 1, multiplier.isFinite else {
            throw PayrollEngineErrorV2.invalidMultiplier
        }
    }

    private static func unique(_ values: [String]) -> [String] {
        var seen = Set<String>()
        return values.filter { seen.insert($0).inserted }
    }
}
